modules/csg_2d/editor/csg_editor_plugin_2d.cpp

```diff
@@ -0,0 +1,804 @@
+					if (xform.xform(handles[i]).distance_to(gpoint) < grab_threshold) {
+						edit_handle = i;
+
+						break;
+					}
+				}
+
+				if (edit_handle == -1) {
+					pressed = false;
+
+					return false;
+				}
+
+				original_mouse_pos = gpoint;
+				original_point = handles[edit_handle];
+				original = get_handle_value(edit_handle);
+				original_transform = node->get_global_transform();
+				last_point = original;
+				pressed = true;
+
+				return true;
+
+			} else {
+				if (pressed) {
+					if (original_mouse_pos != gpoint) {
+						commit_handle(edit_handle, original);
+					}
+
+					edit_handle = -1;
+					pressed = false;
+
+					return true;
+				}
+			}
+		}
+
+		return false;
+	}
+
+	Ref<InputEventMouseMotion> mm = p_event;
+
+	if (mm.is_valid()) {
+		if (edit_handle == -1 || !pressed) {
+			return false;
+		}
+
+		Vector2 cpoint = canvas_item_editor->snap_point(canvas_item_editor->get_canvas_transform().affine_inverse().xform(mm->get_position()));
+		cpoint = node->get_viewport()->get_popup_base_transform().affine_inverse().xform(cpoint);
+		cpoint = original_transform.affine_inverse().xform(cpoint);
+		last_point = cpoint;
+
+		set_handle(edit_handle, cpoint);
+
+		return true;
+	}
+
+	Ref<InputEventKey> k = p_event;
+
+	if (k.is_valid()) {
+		if (edit_handle == -1 || !pressed || k->is_echo()) {
+			return false;
+		}
+
+		if (shape_type == CSGShapeType::RECTANGLE_SHAPE && k->get_keycode() == Key::ALT) {
+			set_handle(edit_handle, last_point); // Update handle when Alt key is toggled.
+		}
+	}
+
+	return false;
+}
+
+void CSGShape2DEditor::forward_canvas_draw_over_viewport(Control *p_overlay) {
+	if (!node) {
+		return;
+	}
+
+	if (!node->is_visible_in_tree()) {
+		return;
+	}
+
+	if (shape_type == CSGShapeType::NONE) {
+		return;
+	}
+
+	if (shape_type == CSGShapeType::POLYGON_SHAPE) {
+		polygon_editor->forward_canvas_draw_over_viewport(p_overlay);
+		return;
+	}
+
+	Transform2D gt = canvas_item_editor->get_canvas_transform() * node->get_screen_transform();
+
+	Ref<Texture2D> h = get_editor_theme_icon(SNAME("EditorHandle"));
+	Vector2 size = h->get_size() * 0.5;
+
+	handles.clear();
+
+	switch (shape_type) {
+		case CSGShapeType::CAPSULE_SHAPE: {
+			CSGCapsule2D *shape = Object::cast_to<CSGCapsule2D>(node);
+			ERR_FAIL_NULL(shape);
+
+			handles.resize(2);
+			float radius = shape->get_radius();
+			float height = shape->get_height() / 2;
+
+			handles[0] = Point2(radius, 0);
+			handles[1] = Point2(0, height);
+
+			p_overlay->draw_texture(h, gt.xform(handles[0]) - size);
+			p_overlay->draw_texture(h, gt.xform(handles[1]) - size);
+
+		} break;
+
+		case CSGShapeType::CIRCLE_SHAPE: {
+			CSGCircle2D *shape = Object::cast_to<CSGCircle2D>(node);
+			ERR_FAIL_NULL(shape);
+
+			handles.resize(1);
+			handles[0] = Point2(shape->get_radius(), 0);
+
+			p_overlay->draw_texture(h, gt.xform(handles[0]) - size);
+
+		} break;
+
+		case CSGShapeType::MESH_SHAPE: {
+			// CSGMesh2D *shape = Object::cast_to<CSGMesh2D>(node);
```

Comment:
Add a `FIXME` or `TODO` comment?

Reply:
There is not really anything to do here as that `CSGShapeType::MESH_SHAPE` has no editor overlay.